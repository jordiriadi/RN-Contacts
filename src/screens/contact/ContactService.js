import Sql from '../../Sql/client';
import Contact from './types/Contact';

export function getContacts() {
  const extractPayload = (payload) => payload.data.map(Contact);

  return Sql.query({}, { method: 'GET' }).then(extractPayload);
}

export function addContact(formValues) {

  return Sql.query({}, { method: 'POST', query: formValues});
}

export function updateContact(id, _formValues) {
  const formValues = {
    firstName: _formValues.firstName,
    lastName: _formValues.lastName,
    age: _formValues.age
  };
  const stringifyParams = `/${id}`;

  return Sql.query({}, { method: 'PUT', query: formValues, stringifyParams})
}

export function deleteContact(id) {
  const stringifyParams = `/${id}`;

  return Sql.query({}, { method: 'DELETE', stringifyParams});
}

export default {
  getContacts,
  addContact,
  updateContact,
  deleteContact
}