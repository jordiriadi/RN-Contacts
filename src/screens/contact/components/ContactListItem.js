import React from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, View, Text, TouchableOpacity } from 'react-native';
import { Avatar } from 'react-native-elements';
import Styles from '../../../utils/Styles';

const ContactListItem = (props) => {
  const {contact, onPress} = props;
 
  return (
    <TouchableOpacity key={contact.id} style={styles.container} onPress={() => onPress(contact)}>
      <View style={styles.avatarContainer}>
        <Avatar
          rounded
          size="medium"
          source={{
            uri:contact.photo
          }}
          title={contact.initial}
        />
      </View>
      <View style={styles.textContainer}>
        <Text style={Styles.DefaultText}>
          {contact.name}
        </Text>
      </View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: Styles.DefaultPadding,
    flexDirection: 'row',
  },
  avatarContainer: {
    flex: 1,
    alignItems: 'center',
  },
  textContainer: {
    flex: 5,
    justifyContent: 'center',
    borderBottomColor: Styles.PrimaryColor,
    borderBottomWidth: 1
  }
});

ContactListItem.propTypes = {
  contact: PropTypes.object,
  onPress: PropTypes.func,
}


export default ContactListItem;