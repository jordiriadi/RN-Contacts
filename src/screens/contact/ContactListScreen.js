import React from 'react';
import { View, SectionList, ActivityIndicator, Text, StyleSheet } from 'react-native';

import ContactListItem from './components/ContactListItem';
import generateSections from '../../utils/generateSections';
import ContactService from './ContactService';
import Styles from '../../utils/Styles';

class ContactListScreen extends React.Component {
  constructor(props){
    super(props);

    this.state = {
      contacts: [],
      loading: false,
    }
  }

  async componentDidMount() {
    this.focusListener = this.props.navigation.addListener('didFocus', () => {
      this.getContacts();
    })
  }

  componentWillUnmount() {
    this.focusListener.remove();
  }

  getContacts = () => {
    console.log('fetch contact ...');
    const _getContacts = ContactService.getContacts();

    const startLoading = () => this.setState({ loading: true });
    const endLoading = () => this.setState({ loading: false });
    
    const handleSuccess = (contacts => {
      const contactsBySections = generateSections(contacts);
      this.setState({
        contacts: contactsBySections,
        loading: false,
      });
    });
    const handleFail = (err) => console.log(err);

    startLoading();
    _getContacts.then(handleSuccess, endLoading).catch(handleFail);

    return _getContacts;
  }

  handleItemPressed = (contact) => this.props.navigation.push('ContactDetailScreen', {contact});

  renderContactItem = ({item}) => <ContactListItem contact={item} onPress={this.handleItemPressed}/>

  renderSectionHeader = ({section}) => (
    <View style={styles.sectionHeader}>
      <Text >{section.title}</Text>
    </View>
  )

  render() {
    const { contacts } = this.state;
    
    return (
      <View>
        {this.state.loading && (
          <ActivityIndicator id="loading-indicator" size="large"/>
        )}
        <SectionList
          id="section-list"
          sections={contacts}
          renderItem={this.renderContactItem}
          renderSectionHeader={this.renderSectionHeader}
          keyExtractor={(item, index) => item + index}
        />
      </View>
    );
  };
}

const styles = StyleSheet.create({
  sectionHeader: {
    paddingLeft: Styles.DefaultPadding,
    paddingTop: Styles.DefaultPadding
  }
})

export default ContactListScreen;