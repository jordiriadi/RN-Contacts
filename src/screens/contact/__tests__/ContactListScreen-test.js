import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';

import ContactListScreen from '../ContactListScreen';
import ContactService from '../ContactService';
import ContactListItem from '../components/ContactListItem';

const mockData = [
  {
    "age": "20",
    "firstName": "Luke",
    "id": "b3abd640-c92b-11e8-b02f-cbfa15db428b",
    "initial": "LS",
    "lastName": "Skywalker",
    "name": "Luke Skywalker",
    "photo": "N/A"
  },
  {
    "age": "33",
    "firstName": "LukeTheSecond",
    "id": "b3abd640-b22g-11e8-b02f-cbfa15db428b",
    "initial": "LS",
    "lastName": "Skywalker",
    "name": "LukeTheSecond Skywalker",
    "photo": "N/A"
  }
]

ContactService.getContacts = jest.fn(() => Promise.resolve(mockData));


describe('Contact List Screen', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<ContactListScreen />);

    expect(wrapper).toMatchSnapshot();
  });

  it('Show Contact List', () => {
    const wrapper = shallow(<ContactListScreen />);

    wrapper.setState({ loading: true });
    expect(wrapper.state().loading).toEqual(true);
    expect(wrapper.find('#loading-indicator'))
  });

  it('Show Contact List', () => {
    const wrapper = shallow(<ContactListScreen />);

    expect(wrapper.find('#section-list'))
  })
});

