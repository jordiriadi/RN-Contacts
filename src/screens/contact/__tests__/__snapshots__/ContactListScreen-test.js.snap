// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact List Screen renders correctly 1`] = `ShallowWrapper {}`;
